File: epubviewer/ui/status_bar.py
from __future__ import annotations

from PySide6.QtWidgets import QLabel, QStatusBar


class ReaderStatusBar(QStatusBar):
    """Bottom bar: chapter title, page-in-chapter progress, and overall book percentage."""

    def __init__(self, parent=None):
        super().__init__(parent)
        self.chapter_label = QLabel("No book loaded")
        self.page_label = QLabel("")
        self.percent_label = QLabel("")
        self.addWidget(self.chapter_label, 1)
        self.addPermanentWidget(self.page_label)
        self.addPermanentWidget(self.percent_label)

    def update_status(self, chapter_title: str, page_index: int, page_count: int, percentage: float) -> None:
        self.chapter_label.setText(chapter_title)
        self.page_label.setText(f"Page {page_index + 1} / {page_count}")
        self.percent_label.setText(f"{percentage:.1f}%")

    def clear_status(self) -> None:
        self.chapter_label.setText("No book loaded")
        self.page_label.setText("")
        self.percent_label.setText("")

File: lib/ui/status_bar.dart
import 'package:flutter/material.dart';

/// Bottom status bar: chapter title, page N/total in chapter, overall percentage.
/// Port of `epubviewer/ui/status_bar.py` (see backup/pyside6/).
class ReaderStatusBar extends StatelessWidget {
  final String chapterTitle;
  final int pageIndex;
  final int pageCount;
  final double percentage;
  final bool hasBook;

  const ReaderStatusBar({
    super.key,
    required this.chapterTitle,
    required this.pageIndex,
    required this.pageCount,
    required this.percentage,
    required this.hasBook,
  });

  @override
  Widget build(BuildContext context) {
    final style = Theme.of(context).textTheme.bodySmall;
    return Container(
      // The decoration (fill + top border) spans the full bar including the area behind
      // the Android system navigation bar; SafeArea keeps the text above that inset so the
      // status bar never overlaps the system gesture/button bar.
      decoration: BoxDecoration(
        color: Theme.of(context).colorScheme.surfaceContainerHighest,
        border: Border(top: BorderSide(color: Theme.of(context).dividerColor)),
      ),
      child: SafeArea(
        top: false,
        child: Container(
          height: 28,
          padding: const EdgeInsets.symmetric(horizontal: 12),
          child: Row(
            children: [
              Expanded(
                child: Text(
                  hasBook ? chapterTitle : 'No book loaded',
                  overflow: TextOverflow.ellipsis,
                  style: style,
                ),
              ),
              if (hasBook) ...[
                Text('Page ${pageIndex + 1} / $pageCount', style: style),
                const SizedBox(width: 16),
                Text('${percentage.toStringAsFixed(1)}%', style: style),
              ],
            ],
          ),
        ),
      ),
    );
  }
}
